README.md
# caldo

A fun javascript game - combine words to win!

## Usage

FIXME

## License

GNU GENERAL PUBLIC LICENSE 
Version 3, 29 June 2007



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-05-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-05-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/caldo/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/caldo/compare/0.1.0...0.1.1


test/caldo/core_test.clj
(ns caldo.core-test
  (:require [clojure.test :refer :all]
            [caldo.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject caldo "0.1.0-SNAPSHOT"
  :description "the hottest new game"
  :url "https://caldo.verbcoach.com"
  :license {:name "GPL 3"
            :url "http://gnu.org/"}
  :dependencies [[babel "1.2.2"]
                 [dag_unify "1.1.0"]
                 [digest "1.4.4"]
                 [environ "1.0.0"]
                 [friend-oauth2 "0.1.3"]
                 [log4j/log4j "1.2.16" :exclusions [javax.mail/mail
                                                    javax.jms/jms
                                                    com.sun.jdmk/jmxtools
                                                    com.sun.jmx/jmxri]]
                 [http-kit "2.1.16"]
                 [org.apache.httpcomponents/httpclient "4.3.5"]
                 [org.clojure/clojure "1.7.0"]
                 [org.clojure/data.json "0.2.5"]
                 [org.clojure/tools.logging "0.2.6"]
                 [org.clojure/tools.nrepl "0.2.7"]
                 [org.postgresql/postgresql "9.4.1208.jre7"]]


  :plugins [[cider/cider-nrepl "0.11.0"]
            [lein-environ "1.0.0"]
            [lein-ring "0.9.7"]
            [s3-wagon-private "1.2.0"]]
  :repositories {"s3" {:url "s3p://ekoontz-repo/releases/"
                       :username :env/aws_access_key ;; gets environment variable AWS_ACCESS_KEY
                       :passphrase :env/aws_secret_key ;; gets environment variable AWS_SECRET_KEY
                       }}
                 
  :resource-paths ["resources"]
  
  ;; caldo.core/app is defined in src/caldo/core.clj.
  :ring {:handler caldo.caldo/app})






src/caldo/core.clj
(ns caldo.core
  (:use
   [hiccup core page]
   [ring.util.codec :as codec])
  (:require
   [caldo.auth.google :as google-auth]
   [cemerick.friend :as friend]
   [cemerick.friend [workflows :as workflows]]
   [clojure.java.io :as io]
   [clojure.tools.logging :as log]
   [compojure.core :refer [context defroutes GET PUT POST DELETE ANY]]
   [compojure.route :as route]
   [compojure.handler :as handler]
   [environ.core :refer [env]]
   [friend-oauth2.workflow :as oauth2]
   [hiccup.page :as h]
   [ring.adapter.jetty :as jetty]
   [ring.middleware.session :as session]
   [ring.middleware.session.cookie :as cookie]
   [ring.middleware.stacktrace :as trace]
   [ring.util.codec :as codec]
   [ring.util.response :as resp]))

(defn foo
  "I don't do a whole lot."
  [x]
  (println x "Hello, World!"))

(defroutes main-routes
  (route/resources "/")

  (GET "/" request
       {:status 200
        :headers {"Content-type" "text/html;charset=utf-8"}
        :body (html [:html [:head [:title "benvenuto a caldo!"]
                            [:link {:rel "stylesheet"
                                    :type "text/css"
                                    :href "/css/caldo.css"}]]
                            
                     [:body
                      [:div#left
                       [:div#header [:h1 "caldo!"]]
                       ]

                      [:div#gamecontainer " "]
                       [:div#userinput
                        [:input {:name "userinput" :size "8"}]]
                       
                      [:script {:type "text/javascript"
                                :src "/js/phaser.min.js"}]
                      [:script {:type "text/javascript"
                                :src "/js/caldo.js"}]]])}))

(def app
  (handler/site 
   (friend/authenticate
    main-routes

    {:allow-anon? true
     :login-uri "/login"
     :default-landing-uri "/"
     :unauthorized-handler #(-> 
                             {:status 401
                              :headers {"Content-type" "text/plain;charset=utf-8"}
                              :body "Unauthorized."})

     :workflows [(workflows/interactive-form)
                 (oauth2/workflow google-auth/auth-config)]})))




src/caldo/caldo.clj
(ns caldo.caldo
  (:refer-clojure :exclude [get-in])
  (:require
   [dag_unify.core :refer [get-in unify]]
   [babel.italiano :refer [generate parse]]
   [caldo.auth.google :as google-auth]
   [cemerick.friend :as friend]
   [cemerick.friend [workflows :as workflows]]
   [clojure.data.json :refer [write-str]]
   [clojure.java.io :as io]
   [clojure.string :as string]
   [clojure.tools.logging :as log]
   [compojure.core :refer [context defroutes GET PUT POST DELETE ANY]]
   [compojure.route :as route]
   [compojure.handler :as handler]
   [environ.core :refer [env]]
   [friend-oauth2.workflow :as oauth2]
   [hiccup.core :as h :refer [html]]
   [ring.adapter.jetty :as jetty]
   [ring.middleware.session :as session]
   [ring.middleware.session.cookie :as cookie]
   [ring.middleware.stacktrace :as trace]
   [ring.util.codec :as codec]
   [ring.util.response :as resp]))

(declare get-roots)

(defroutes main-routes
  (route/resources "/")

  (GET "/" request
       {:status 200
        :headers {"Content-type" "text/html;charset=utf-8"}
        :body (html [:html [:head [:title "benvenuto a caldo!"]
                            [:link {:rel "stylesheet"
                                    :type "text/css"
                                    :href "/css/caldo.css"}]
                            [:script {:type "text/javascript"
                                      :src "/js/jquery.min.js"}]
                            [:script {:type "text/javascript"
                                      :src "/js/mustache.min.js"}]
                            [:script {:type "text/javascript"
                                      :src "/js/phaser.min.js"}]
                            [:script {:type "text/javascript"
                                      :src "/js/log4.js"}]
                            [:script {:type "text/javascript"
                                      :src "/js/caldo.js"}]
                            ]
                     ;; See ../../resources/public/js/caldo.js for definition of
                     ;; the caldo() onload function: caldo().
                     ;; See ../../resources/public/mst/caldo.mst for HTML template
                     ;; used by caldo().
                     [:body {:onload "caldo();"} ]])})

  (GET "/say" request
       {:status 200
        :headers {"Content-type" "application/json;charset=utf-8"}
        :body (let [expr (:expr (:params request))
                    parsed (reduce concat (map :parses (parse expr)))
                    roots (get-roots parsed)
                    debug (log/debug (str "# parses('" expr "'): " (count parsed)))]
                (log/info (str "roots " (string/join ";" roots)))
                (write-str {:roots roots}))}))

(def app
  (handler/site 
   (friend/authenticate
    main-routes

    {:allow-anon? true
     :login-uri "/login"
     :default-landing-uri "/"
     :unauthorized-handler #(-> 
                             {:status 401
                              :headers {"Content-type" "text/plain;charset=utf-8"}
                              :body "Unauthorized."})

     :workflows [(workflows/interactive-form)
                 (oauth2/workflow google-auth/auth-config)]})))

(defn get-roots [parses]
  (set
   (mapcat (fn [parse]
             [(get-in parse [:comp :italiano :italiano])
              (if (get-in parse [:root :italiano :infinitive])
                (get-in parse [:root :italiano :infinitive])
                (get-in parse [:root :italiano :italiano]))])
           parses)))


src/caldo/auth/google.clj
(ns caldo.auth.google
  (:require [cemerick.friend [workflows :as workflows]]
            [cemerick.friend :as friend]
            [clojure.data.json :as json]
            [clojure.tools.logging :as log]
            [compojure.core :as compojure :refer [ANY GET]]
            [compojure.handler :as handler]
            [compojure.route :as route]
            [friend-oauth2.workflow :as oauth2]
            [friend-oauth2.util :refer [format-config-uri]]
            [org.httpkit.client :as http]
            [korma.core :as k]))

(require '[environ.core :refer [env]])

(derive ::admin ::user)

(declare insert-session)

(def client-config
  {:client-id (env :google-client-id)
   :client-secret (env :google-client-secret)
   :callback {:domain (env :google-callback-domain)
              :path "/oauth2callback"}})

(if-let [google-client-id (env :google-client-id)]
  (log/info (str "google authentication configured client-id:" google-client-id)))

(if-let [google-callback-domain (env :google-callback-domain)]
  (log/info (str "google callback-domain: " google-callback-domain)))

(defn credential-fn [token]
  ;;lookup token in DB or whatever to fetch appropriate :roles
  {:identity token :roles #{::user}})

(def uri-config
  {:authentication-uri {:url "https://accounts.google.com/o/oauth2/auth"
                        :query {:client_id (:client-id client-config)
                                :response_type "code"
                                :redirect_uri (format-config-uri client-config)
                                :scope "email"}}

   :access-token-uri {:url "https://accounts.google.com/o/oauth2/token"
                      :query {:client_id (:client-id client-config)
                              :client_secret (:client-secret client-config)
                              :grant_type "authorization_code"
                              :redirect_uri (format-config-uri client-config)}}})

(def auth-config {:client-config client-config
                  :uri-config uri-config
                  :credential-fn credential-fn})

;; TODO: should be one level above this (verbcoach.auth)
(defn is-authenticated [if-authenticated]
  (if (not (nil? (friend/current-authentication)))
    if-authenticated
     (do (log/debug (str "is-authenticated: not authenticated; redirecting to /"))
         {:status 302
          :headers {"Location" "/"}})))

(defn request2cookie [request]
  "get the cookie (if any) embedded at a certain path within the request params."
  (if (and request
           (:cookies request)
           (get (:cookies request) "ring-session"))
    (:value (get (:cookies request) "ring-session"))))

(defn insert-session-if-none
  ([session]
     (if (nil? session)
       (log/warn (str "session was unexpectedly null."))

       (let [session-row (first (k/exec-raw ["SELECT session.*
                                                FROM session 
                                               WHERE ring_session = ?::uuid" [session]] :results))
             debug (log/trace (str "insert-session-if-none: ring-session:" session))
             ]
         (if (not session-row)
           (insert-session session)))))
  
  ([session access-token user-id]
     (let [sql-only (log/debug (str "SQL ONLY: "
                                   (k/sql-only ["SELECT session.*
                                                   FROM session 
                                                  WHERE ring_session = ?::uuid
                                                    AND access_token = ?
                                                    AND user_id = ?" [session access-token user-id]])))
           session-row (first (k/exec-raw ["SELECT session.*
                                              FROM session 
                                             WHERE ring_session = ?::uuid
                                               AND access_token = ?
                                               AND user_id = ?" [session access-token user-id]] :results))
           debug (log/trace (str "insert-session-if-none: session:" session))
           ]
       (if (not session-row)
         (insert-session user-id access-token session)))))

;; TODO: factor out update/insert into separate function with more representative names
(defn token2username [access-token & [request]]
  "Get user's email address from the vc_user table, given their access token. 
   If access-token is not registered in database,
   ask Google for the user's email address and add as new row to the vc_user table."
  (cond
   (nil? access-token)
   (do
     (log/error "token2user: supplied access-token was null.")
     (throw (Exception. (str "token2user: supplied access-token was null."))))

   true
   (let [debug (log/debug (str "token2username: access-token: " access-token))
         user-by-access-token
         (first (k/exec-raw [(str "SELECT email,session.user_id AS userid
                                    FROM vc_user 
                              INNER JOIN session
                                      ON (vc_user.id = session.user_id) 
                                     AND session.access_token=?")
                             [access-token]] :results))]
     (if user-by-access-token
       (let [email (:email user-by-access-token)
             user-id (:userid user-by-access-token)
             ring-session (get-in request [:cookies "ring-session" :value])]
         
         (log/debug (str "found user by access-token in Postgres vc_user database: email: "
                        email))
         (if request (do
                       (log/trace (str "found user by access-token, and currently the request is: " request))
                       (log/debug (str "found user by access-token, and currently the ring-session (from request) is: " ring-session))
                       (insert-session-if-none (get-in request [:cookies "ring-session" :value]) access-token user-id)))
         email)

       ;; else, user could not be found locally by searching the 'session' table access token, so get user's info from google by using the access-token.
       (do
         (log/info (str "user's token was not in database: querying: https://www.googleapis.com/oauth2/v1/userinfo?access_token=<input access token> to obtain user info"))
         (let [{:keys [status headers body error] :as resp} 
               @(http/get 
                 (str "https://www.googleapis.com/oauth2/v1/userinfo?access_token=" access-token))]
           (cond error
                 (log/warn "Failed, exception: " error)
                 
                 (not (= status 200))
                 (do
                   (log/error (str "Received a non-200 status from Google: status=" status))
                   (log/error (str "Received a non-200 status from Google: body=" body))
                   nil)
                 
                 true
                 (do
                   (log/debug (str "body: " body))
                   (let [body (json/read-str body
                                             :key-fn keyword
                                             :value-fn (fn [k v]
                                                         v))
                         email (get body :email)
                         given-name (get body :given_name)
                         family-name (get body :family_name)
                         picture (get body :picture)]
                     (log/debug (str "Google says user's email is: " email))
                     (log/debug (str "Google says user's given_name is: " given-name))
                     (log/debug (str "Google says user's family_name is: " family-name))
                     (log/debug (str "Google says user's picture is: " picture))

                     ;; insert a new user record if necessary.
                     (let [user-id (:id (first (k/exec-raw [(str "SELECT id 
                                                                   FROM vc_user 
                                                                  WHERE email=?")
                                                            [email]] :results)))
                           user-id (if user-id
                                     (do
                                       (log/info (str "found existing user record with email=" email ";user-id=" user-id))
                                       user-id)
                                    
                                     (do
                                       (log/info (str "inserting new user record."))
                                       (:id
                                        (first
                                         (k/exec-raw [(str "INSERT INTO vc_user (given_name,family_name,picture,updated,email)
                                                        VALUES (?,?,?,now(),?) RETURNING id")
                                                      [given-name family-name picture email]] :results)))))]
              
                       ;; insert or update session based on this user_id.
                       (let [session-row (first (k/exec-raw ["SELECT session.*
                                                               FROM vc_user
                                                         INNER JOIN session
                                                                 ON (session.user_id = vc_user.id)
                                                              WHERE vc_user.id=?" [user-id]] :results))

                             ring-session (get-in request [:cookies "ring-session" :value])
                             ]
                         (insert-session-if-none (get-in request [:cookies "ring-session" :value]) access-token user-id))
                                                  
                       email))))))))))

(defn insert-session
  ([ring-session]
     (if (nil? ring-session)
       (log/warn (str "not inserting session since ring_session is nil."))

       ;; else
       (do
         (log/info (str "inserting new session record (session=" ring-session))
         (k/exec-raw [(str "INSERT INTO session (ring_session)
                             VALUES (?::uuid)")
                      [ring-session]]))))

  ([user-id access-token ring-session]
     (if (nil? ring-session)
       (log/warn (str "not inserting session since ring session is nil."))

       ;; else
       (do
         (log/debug (str "inserting new session record (session=" ring-session))
         (log/debug (str "                             (access-token=" access-token))
         (log/debug (str "                             (user-id=" user-id))
         (k/exec-raw [(str "INSERT INTO session (access_token,user_id,ring_session)
                             VALUES (?,?,?::uuid)")
                      [access-token user-id ring-session]])))))

(defn token2info [access-token]
  (first (k/exec-raw [(str "SELECT vc_user.* 
                              FROM vc_user
                        INNER JOIN session
                                ON (vc_user.id = session.user_id)
                             WHERE session.access_token=?")
                      [access-token]]
                     :results)))

(defn token2picture [access-token]
  (let [token2info (token2info access-token)]
    (if token2info
      (:picture token2info))))

(def routes
  (compojure/routes
   (GET "/login" request
        (do
          (log/debug (str "running /login with google-client-id: " (:client-id client-config) "; client request: " request))

          ;; need all three environment variables set correctly; otherwise thrown an exception.
          (if (nil? (:client-id client-config))
            (do
              (log/error (str "No google client id was found in the environment."))
              (throw (Exception. (str "Administrator must define GOOGLE_CLIENT_ID in my runtime environment.")))))

          (if (nil? (:client-secret client-config))
            (do
              (log/error (str "No google client secret was found in the environment."))
              (throw (Exception. (str "Administrator must define GOOGLE_CLIENT_SECRET in my runtime environment.")))))

          (if (nil? (:callback client-config))
            (do
              (log/error (str "No google client callback was found in the environment."))
              (throw (Exception. (str "Administrator must define GOOGLE_CALLBACK_DOMAIN in my runtime environment.")))))


          (friend/authorize #{::user} "Authorized page.")
          
          (is-authenticated
           (do
             (let [username (token2username
                             (-> request :session :cemerick.friend/identity :current :access-token)
                             request)]
               (log/debug (str "Logging in user: " username " with access token: " 
                               (-> request :session :cemerick.friend/identity :current :access-token))))
             {:status 302
              :headers {"Location" "/"}}))))))


resources/public/js/pong.js
// welcome to caldo!
// note that phaser{.min}.js must be loaded for caldo.js to work.

var game =
    new Phaser.Game(
	480, 320,
	Phaser.AUTO,
	null, {
	    preload: preload,
	    create: create,
	    update: update
	});

var ball;
var paddle;

var bricks;
var newBrick;
var brickInfo;
var liveBricks;

var scoreText;
var score = 0;

var lives = 3;
var livesText;
var lifeLostText;

var playing = false;
var startButton;

function preload() {
    game.scale.scaleMode = Phaser.ScaleManager.SHOW_ALL;
    game.scale.pageAlignHorizontally = true;
    game.scale.pageAlignVertically = true;
    game.stage.backgroundColor = "#edc";
    game.load.image('ball','img/ball.png');
    game.load.image('paddle','img/paddle.png');
    game.load.image('brick','img/brick.png');
    game.load.spritesheet('ball','img/wobble.png',20,20);
    game.load.spritesheet('button','img/button.png',120,40);
}

function create() {
    game.physics.startSystem(Phaser.Physics.ARCADE);
    ball = game.add.sprite(50,250,'ball');
    ball.animations.add('wobble', [0,1,0,2,0,1,0,2,0], 24);
    paddle = game.add.sprite(game.world.width*0.5,
			     game.world.height-5,
			     'paddle');
    game.physics.enable(ball,Phaser.Physics.ARCADE);
    game.physics.enable(paddle,Phaser.Physics.ARCADE);
    ball.body.collideWorldBounds = true;
    ball.body.bounce.set(1);
    ball.body.gravity.y = 0;
    game.physics.arcade.checkCollision.down = false;

    ball.checkWorldBounds = true;
    ball.events.onOutOfBounds.add(ballLeaveScreen,this);
    
    paddle.anchor.set(0.5,1);
    paddle.body.immovable = true;

    textStyle = { font: '18px Arial', fill: '#0095DD' };
    
    scoreText = game.add.text(5,5,'Points: 0',textStyle);
    

    brickText = game.add.text(200,5,'Bricks: ',textStyle);

    initBricks();
    
    livesText = game.add.text(game.world.width-5, 5, 'Lives: ' + lives,
			      textStyle);
    livesText.anchor.set(1,0);
    lifeLostText = game.add.text(game.world.width*0.5,
				 game.world.height*0.5,
				 'Life lost, click to continue',
				 textStyle);
    lifeLostText.anchor.set(0.5);
    lifeLostText.visible = false;
    startButton = game.add.button(game.world.width*0.5,
				  game.world.height*0.5,
				  'button',
				  startGame, this,
				  1,0,2);
}

function update() {
    game.physics.arcade.collide(ball,paddle,ballHitPaddle);
    game.physics.arcade.collide(ball,bricks,ballHitBrick);
    if (playing) {
	paddle.x = game.input.x || game.world.width*0.5;
    }
}

function initBricks() {
    liveBricks = 0;
    brickInfo = {
	width: 50,
	height: 20,
	count: {
	    col: 7,
	    row: 3
	},
	offset: {
	    top: 50,
	    left: 60
	},
	padding: 10
    }

    bricks = game.add.group();
    for (c = 0; c < brickInfo.count.col; c++) {
	for (r = 0; r < brickInfo.count.row; r++) {

	    // create new brick and add to the group
	    var brickX =
		(c*(brickInfo.width+brickInfo.padding)) +
		brickInfo.offset.left;
	    var brickY =
		(r*(brickInfo.height+brickInfo.padding)) +
		brickInfo.offset.top;
	    newBrick =
		game.add.sprite(brickX, brickY, 'brick');
	    game.physics.enable(newBrick, Phaser.Physics.ARCADE);
	    newBrick.body.immovable = true;
	    newBrick.anchor.set(0.5);
	    bricks.add(newBrick);
	    liveBricks++;
	}
    }
    brickText.setText('Bricks: ' + liveBricks);
}


function startGame() {
    startButton.destroy()
    ball.body.velocity.set(150,-150);
    playing = true;
}

function ballLeaveScreen() {
    lives--;
    if (lives) {
	livesText.setText('Lives:'+lives);
	lifeLostText.visible = true;
	ball.reset(game.world.width*0.5,game.world.height-25);
	paddle.reset(game.world.width*0.5,game.world.height-5);
	game.input.onDown.addOnce(function() {
	    lifeLostText.visible = false;
	    ball.body.velocity.set(150,-150);
	},this);
    } else {
	alert('You lost, game over!');
	location.reload();
    }
}

function ballHitPaddle(ball,paddle) {
    ball.animations.play('wobble');
    ball.body.velocity.x = -1*5*(paddle.x - ball.x);
}

function ballHitBrick(ball,brick) {
    var killTween = game.add.tween(brick.scale);
    killTween.to({x:0.25,y:0.25},150,Phaser.Easing.Linear.Out,true,10);
    killTween.onComplete.addOnce(function() {
	brick.kill();
    }, this);
    killTween.start();
    ball.animations.play('wobble');
    score += 10;
    scoreText.setText('Points: ' + score);
    liveBricks--;
    brickText.setText('Bricks: ' + liveBricks);
    if (liveBricks == 0) {
	alert('You won the game, congratulations!');
	location.reload();
    }
}


resources/public/js/caldo.js
// welcome to caldo!
// Note that all of:
//
// - jquery{.min}.js
// - mustachephaser{.min}.js
// - phaser{.min}.js 
// - log4.js
//
// must be loaded for caldo.js to work.


// global constants
var num_words_at_a_time = 10;
var num_shelves = 2;
var newWordInterval = [3000,4000];
var logging_level = DEBUG;
var hang_shelves = false;

// TODO: load from server.
var shelf_words = [
    ["io","tu","lui","lei","noi","voi","loro"],
    ["avere","fare","lavorare","leggere","mangiare",
     "studiare","suonare","tornare"],
    ["si","lo","la"]];

// global variables
var words;
var word;
var shelves;
var text;
var game;
var scoreText;
var score = 0;

// methods
function respond_to_user_input(event) {
    key_pressed = event.which;
    if (key_pressed == 13) {
	// send to language server:
	$.ajax({
	    cache: false,
	    type: "GET",
	    data: {expr: $("#userinput").val()},
            dataType: "json",
            url: "/say"}).done(function(content) {
		roots = content.roots;
		if (roots == undefined) {
		    log(INFO,"server returned no roots for input: '" + $("#userinput").val());
		} else {
		    log(INFO,"response from server: found: " + roots.length + " roots.");
		}

		remove_from_blocks(roots);

		$("#userinput").val("");
		$("#userinput").focus();
	    });
    }
}

function remove_from_blocks(roots) {
    num_roots = roots.length;
    num_blocks = words.children.length;
    for (i = 0; i < num_roots; i++) {
	root = roots[i];
	for (c = 0; c < num_blocks; c++) {
	    block = words.children[c];
	    if (block.alive == true) {
		block_text = block._text; // TODO: is this best practices per Phaser docs to access "_" fields?
		if (block_text === root) {
		    kill_block(block);
		    break; // only kill one block that matches the text; otherwise game is too easy.
		}
	    }
	}
    }
}

function kill_block(brick) {
    var killTween = game.add.tween(brick.scale);
    killTween.to({x:0.25,y:0.25},150,Phaser.Easing.Linear.Out,true,10);
    killTween.onComplete.addOnce(function() {
	brick.kill();
    }, this);
    killTween.start();
    score += 10;
    scoreText.setText('Points: ' + score);

}

function caldo() {
    $.get('/mst/caldo.mst', function(template) {
	view = {
	    "caldo": [
		{}
	    ],
	    "caldo": [
		{}
	    ],
	    "name": function () {
		return "caldo";
	    }
	}

	// 1. populate page with HTML:
	output = Mustache.render(template,view);
	$("body").html(output);

	// 2. connect input element to game.
	$("#userinput").keypress(respond_to_user_input);

	// 3. initialize input elements
	$("#userinput").val("");
	$("#userinput").focus();
    
	// 4. start game
	game = new Phaser.Game(
	    600,400,
	    Phaser.AUTO,
	    "gamecontainer", {
		preload: preload,
		create: create,
		update: update
	    });

	// TODO: loop over num_shelves
	window.setInterval(function() {newWord(0);},
			   newWordInterval[0]);
	window.setInterval(function() {newWord(1);},
			   newWordInterval[1]);
    });
}

function preload() {
    game.scale.scaleMode = Phaser.ScaleManager.SHOW_ALL;
    game.scale.pageAlignHorizontally = true;
    game.scale.pageAlignVertically = true;
    game.stage.backgroundColor = "#ede";
    game.load.image('word','img/ball.png');
    game.load.image('shelf','img/brick.png');
}

function create() {
    game.physics.startSystem(Phaser.Physics.ARCADE);

    textStyle = { font: '18px Arial', fill: '#0095DD' };
    scoreText = game.add.text(5,5,'Points: 0',textStyle);
    
    shelves = game.add.group();
    words = game.add.group();
    for (c = 0; c < num_shelves; c++) {
	if (hang_shelves == true) {
	    shelf = game.add.sprite((150*c)+100,game.world.height*0.9,'shelf')
	    game.physics.enable(shelf,Phaser.Physics.ARCADE);
	    shelf.body.gravity.y = 0; // shelf is floating in the air.
	    shelf.body.bounce.set(0.5);
	    shelf.body.collideWorldBounds = true;
	    shelf.body.immovable = true;
	    shelves.add(shelf);
	}
	newWord(c);
    }
}

function wordHitShelf(word,shelf) {
    foo = 32;
}

function update() {
    game.physics.arcade.collide(words,shelves,wordHitShelf);
    game.physics.arcade.collide(words,words,wordHitShelf);
}

function randomWord(shelf) {
    var random_integer = Math.floor(Math.random() * shelf_words[shelf].length);
    return shelf_words[shelf][random_integer];
}

function newWord(shelf) {
    // count the number of alive words: TODO: do this with a global variable.
    alive_words = 0;
    num_blocks = words.children.length;
    for (c = 0; c < num_blocks; c++) {
	block = words.children[c];
	if (block.alive == true) {
	    alive_words++;
	}
    }

    if (alive_words < num_words_at_a_time) {
	style = { font: "32px Arial",
		  fill: "#0055ee",
		  wordWrap: true,
		  align: "center",
		  backgroundColor: "#ffffef" };
	
	text = game.add.text( (150*shelf) + 125, 0, randomWord(shelf), style);
	text.anchor.set(0.5);
	
	game.physics.arcade.enable([text]);
	text.body.velocity.setTo(0,0);
	text.body.gravity.y = 500;
	text.body.collideWorldBounds = true;
	text.body.bounce.set(0.5 + (.1 * (Math.random())));
	words.add(text);
    }
}
    
